Streaming: replace npmlog with pino & pino-http

ThisIsMissEm opened the pull request:
This implements #27387. 

**Advice for reviewers:** I'd recommend going commit by commit in the files changed tab, as each commit tries to focus in on specific changes (though the change from `verifyClient` to `server.on("upgrade")` is bundled with the logger change)

In order to properly initialize the logging, I had to rework the WebSocket Server attachment to the HTTP server to use the upgrade event, which allows for making a decision around authentication before even attempting to upgrade the connection, which is preferrable to `verifyClient` whose usage is now discouraged. 

By using the upgrade event and `ws.handleUpgrade` method allows us to pass through a logger to the `wss.on("connection")` event/handler, which can be augmented with logging details specific to websockets.

The log redaction and sanitizer is a little complicated, but essentially prevents access tokens from accidentally ending up in server logs.

The `request.id` is now managed by the logger, and we don't need `request.remoteAddress` as the request logger correctly captures the information about the request for logging, so we're able to remove both of those middlewares. Additionally, I replaced the custom CORS middleware with the standard and widely used `cors` package, which is well tested.

Bundled in is the start of some code cleanup, moving things out to separate files where appropriate (since these things where making it much more difficult to read the code whilst I changed the connection upgrade handling).

In development, if you want, you can pipe to pino-pretty for nicer looking logs, which is available via a development dependency:

```sh
$ yarn workspace @mastodon/streaming start | pino-pretty
```

<img width="1128" alt="Screenshot 2023-11-12 at 9 17 57 pm" src="https://github.com/mastodon/mastodon/assets/30827/de2d7932-ff19-4b3d-a7a8-489c552f6521">

ThisIsMissEm commented:
@renchap  [9b24627](https://github.com/mastodon/mastodon/pull/27828/commits/9b2462731af4074d30440d008c7881d251b8b851) is what I was talking about over on #24702, it didn't directly break anything when linting ran, but did cause issues; This will also act as a starting point for eventually being able to migrate to ESM, without affecting the reset of the codebase.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -25,15 +29,86 @@ dotenv.config({
+    return req.id;
+  }
+
+  req.id = uuid.v4();
+
+  // Allow for usage with WebSockets:
+  if (res) {
+    res.setHeader('X-Request-Id', req.id);
+  }
+
+  return req.id;
+}
+
+/**
+ * Request log sanitizer to prevent logging access tokens in URLs
+ * @param {http.IncomingMessage} req
+ */
+const sanitizeReqLog = (req) => {
+  const log = pinoHttpSerializers.req(req);
+  if (typeof log.url === 'string') {
+    log.url = log.url.replace(/(access_token)=([A-Z0-9a-z_]+)/gi, '$1=[Redacted]');
+  }
+  return log;
+};
+
+const httpLogger = pinoHttp({
+  logger,
+  genReqId,
+  serializers: {
+    req: sanitizeReqLog
+  }
+});
+
+if (process.env.LOG_LEVEL && Object.keys(logger.levels.values).includes(process.env.LOG_LEVEL)) {
+  logger.level = process.env.LOG_LEVEL;
+} else if (environment === 'development') {
+  logger.level = 'debug';
+} else {
+  logger.level = 'info';
+}
```
I am tempted to move this into a generic `setupLogging` function that takes in the environment, and returns a `logger` and `httpLogger` back and also exposes the `sanitizeRequestForLogging` (`sanitizeReqLog`) method.

Would appreciate thoughts on if this makes sense as a change?

mjankowski commented:
This is definitely easier to process one commit at a time than all together, but even so, it seems like at least:

- The eslint/tsconfig changes
- Extracting the metrics to a separate file
- Move isTruthy to utils
- Swap out custom impl for `cors` package

...could all be stand-alone PRs (they all seem like improvements), and may make it easier to review/merge the narrower stated purpose here of the logger change?

ThisIsMissEm commented:
@mjankowski I've tried in the past to split changes down, but due to the project merge cadence I had to abandon those PRs which were all stacked and a nightmare to maintain / rebase.

I'm trying to avoid that here.

ThisIsMissEm commented:
Extracted the CORS change to https://github.com/mastodon/mastodon/pull/28523/ & rebased this PR to separate the server initialization changes from the logging changes as best as possible.

ThisIsMissEm commented on streaming/logging.js:
```diff
@@ -0,0 +1,98 @@
+function sanitizeRequestLog(req) {
+  const log = pinoHttpSerializers.req(req);
+  if (typeof log.url === 'string' && log.url.includes('access_token')) {
+    log.url = log.url.replace(/(access_token)=([A-Z0-9a-z_-]+)/gi, '$1=[Redacted]');
```
@ClearlyClaire could you confirm that access tokens conform to this regex? I believe they should, but I'm not 100% sure.

ClearlyClaire replied:
As far as I understand Devise's code, this uses `SecureRandom.urlsafe_base64` without padding, so this looks accurate!

renchap commented:
Now that the CORS refactor is merged, could you extract your websocket refactor commit into it's own PR?

Then I think we will be able to review & merge this :)

Thanks a lot for this work!

ThisIsMissEm commented:
> Now that the CORS refactor is merged, could you extract your websocket refactor commit into it's own PR?
> 
> Then I think we will be able to review & merge this :)
> 
> Thanks a lot for this work!

With all due respect, this is the fourth time I've done the work in this branch & tried to get it merged. I've had to abandon 3 other attempts. My time, like yours is extremely limited. I have tried to make the changes reviewable commit by commit: https://github.com/mastodon/mastodon/pull/27828/commits/6533a523d37e5cf5afb7ec012409f141f2064e33 & https://github.com/mastodon/mastodon/pull/27828/commits/93351e1c6c865c281687dfd190720246de58b8bb

I do not have time to continuously rebase this work, nor do I have the energy as I'm currently reasonably unwell. You either want this change and for the streaming server to improve, or you don't. I know your resources are thin, but mine are likely thinner. I've been trying to get this work merged for almost a year at this point.

ThisIsMissEm commented:
There is a follow-on related to this pull request here: https://github.com/mastodon/mastodon/pull/28632

ThisIsMissEm committed:
Streaming: Refactor to use Pino & Pino-Http for logging

We needed to rework certain method signatures to allow passing the logger in, rather than relying on req.log which may not be defined for websocket connections. The Request ID is now handled by the logger, so that custom middleware has been removed, and the RemoteAddress middleware is no longer necessary with structured logging where each message carries the request ID.

ThisIsMissEm committed:
Streaming: Fix garbage collection code

By setting the properties to null, instead of just triggering a `delete` on the entire session object, we ran into typechecking issues.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -28,15 +27,18 @@ dotenv.config({
 const createRedisClient = async (config) => {
   const { redisParams, redisUrl } = config;
+  // @ts-ignore
   const client = new Redis(redisUrl, redisParams);
-  client.on('error', (err) => log.error('Redis Client Error!', err));
+  // @ts-ignore
+  client.on('error', (err) => logger.error({ err }, 'Redis Client Error!'));
```
These `@ts-ignore`s are due to a typing issue with the `err` parameter here.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -61,12 +63,12 @@ const parseJSON = (json, req) => {
     if (req) {
       if (req.accountId) {
-        log.warn(req.requestId, `Error parsing message from user ${req.accountId}: ${err}`);
+        req.log.error({ err }, `Error parsing message from user ${req.accountId}`);
       } else {
-        log.silly(req.requestId, `Error parsing message from ${req.remoteAddress}: ${err}`);
+        req.log.error({ err }, `Error parsing message from ${req.remoteAddress}`);
       }
     } else {
-      log.warn(`Error parsing message from redis: ${err}`);
+      logger.error({ err }, `Error parsing message from redis`);
     }
     return null;
   }
```
I'd ideally like to remove this mess, and instead handle the `parseJSON` method as a Result object `{ error, value }`, which would allow us to pass the error back to the call-site for logging, rather than performing the logging in this method.

I'm unsure if I should make this change now or not.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -195,26 +207,33 @@ const startServer = async () => {
 
   // Handle upgrade requests:
   server.on('upgrade', async function handleUpgrade(request, socket, head) {
+    const startTime = Date.now();
+
```
This is only used for the responseTime logging, which I'm not sure if we really need here, so we could remove this and line 267.

ClearlyClaire replied:
I don't feel strongly either way. I think this is not useful in most of cases, but it could be in some?

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -238,15 +257,34 @@ const startServer = async () => {
 
+      request.log.info("Authenticated request & upgraded to WebSocket connection");
+
+      const wsLogger = logger.child({
```
The logger child that we're creating here is explicitly a child of the root logger and not `request.log` as the former would include all the request properties with every websocket log message, which would be excessive. Instead we're just passing through `request.id` for correlation purposes.

I could perhaps rename `websocketLogger` to `websocketRequestLogger` and have a `createWebsocketLogger(request)` here.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -238,15 +257,34 @@ const startServer = async () => {
+
       // Start the connection:
-      wss.emit('connection', ws, request);
+      wss.emit('connection', ws, request, wsLogger);
```
For now I'm adding a third argument to `onConnection`, but in the future, we may be able to attach the "logger" directly to the `websocket` (`ws`)

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -273,8 +311,10 @@ const startServer = async () => {
   // When checking metrics in the browser, the favicon is requested this
   // prevents the request from falling through to the API Router, which would
   // error for this endpoint:
+  // @ts-ignore
   app.get('/favicon.ico', (req, res) => res.status(404).end());
 
+  // @ts-ignore
   app.get('/api/v1/streaming/health', (req, res) => {
```
These were due to unused parameters, we may not need those `@ts-ignore`s anymore.

ThisIsMissEm commented:
@ClearlyClaire I'm just refactoring a bit of this into a separate PR which tackles the type issues with Redis & Postgresql (I'm going to extract these things to separate files since the fixes ended up being kind of messy)

ThisIsMissEm committed:
Streaming: Fix various minor typechecking errors

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -441,6 +444,7 @@ const startServer = async () => {
         return;
       }
 
+      // @ts-ignore
```
This is because the types for `pg` don't declare it as a union on `Pool.connect` of `(err: Error, client: never, done: (release?: any) => void) | (err: never, client: PoolClient, done: (release?: any) => void)` which would allow discrimination of the types, so instead you get an error of "client may be undefined" for `client.query`

We could alternatively use async/await here, which would side-step that issue of `Pool.connect`'s types being incorrect since: `const client = await pgPool.connect()` is correctly typed as `Promise<PoolClient>;`

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -451,6 +455,7 @@ const startServer = async () => {
 
         if (result.rows.length === 0) {
           err = new Error('Invalid access token');
+          // @ts-ignore
           err.status = 401;
```
This is being refactored separately, as I've a branch that introduces proper Error objects for the various errors we throw in the code.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -577,19 +585,23 @@ const startServer = async () => {
-   * @returns {function(object): void}
+   * @returns {SubscriptionListener}
    */
   const createSystemMessageListener = (req, eventHandlers) => {
```
I have a separate commit that completely refactors this code, so for now I've just added a minor guard to ensure messages from ruby over redis to streaming are well-formed.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -390,30 +416,7 @@ const startServer = async () => {
-
-  /**
-   * @param {any} req
+   * @param {http.IncomingMessage & {scopes: string[]}} req
```
I could also change this to 
```suggestion
   * @param {http.IncomingMessage & ResolvedAccount} req
```

As I have done so elsewhere.

ThisIsMissEm commented:
@ClearlyClaire @mjankowski okay, I think this should be good to go; I've moved the postgresql and redis type issues into a separate PR (still to open / finish, because I want to move those out of `index.js`)

ClearlyClaire commented:
Apart from the minor naming issues, this looks fine to me, but I'd like another set of eyes on this.

Also, this is a (fairly minor) breaking change that needs to be documented.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -195,39 +198,45 @@ const startServer = async () => {
   server.on('upgrade', async function handleUpgrade(request, socket, head) {
+    const startTime = Date.now();
+
+    websocketLogger(request);
```
Have updated this to `attachWebsocketHttpLogger` and created a `createWebsocketLogger` for the actual logger below in the handleUpgrade.

ThisIsMissEm commented:
> Also, this is a (fairly minor) breaking change that needs to be documented.

@ClearlyClaire currently the documentation site has absolutely zero mention of logging, and only a little bit on monitoring Mastodon. I did add some documentation about streaming's metrics to the "scaling up" docs, but I don't see a good place for the notice that this is a breaking change in how the streaming server produces logs.

By the looks the changelog should be updated in the release branch.

ClearlyClaire commented:
> > Also, this is a (fairly minor) breaking change that needs to be documented.
> 
> @ClearlyClaire currently the documentation site has absolutely zero mention of logging, and only a little bit on monitoring Mastodon. I did add some documentation about streaming's metrics to the "scaling up" docs, but I don't see a good place for the notice that this is a breaking change in how the streaming server produces logs.
> 
> By the looks the changelog should be updated in the release branch.

Yes, I meant that as a note to add something in the release notes when the time comes, not as something to do before the PR gets merged.

renchap reviewed (approved):
Looks good to me. Quite a hard change to test entirely, but it makes the code cleaner and hopefully provides more useful logs.
